### ISSUE TITLE
Resources are leaked if host function overflows the stack

### ISSUE THREAD

**Issue by matklad** [wasmerio/wasmer#2328]
### Describe the bug

If host function overflows the stack, wasmer seems to unwind the stack past it, bypassing destructors, leading to a resource leak. 

### Steps to reproduce

https://github.com/matklad/wasmer-repro

```
$ git clone https://github.com/matklad/wasmer-repro && cd wasmer-repro
$ cargo run --example bench
# wait for OOM to happen
```

The gist of the problem:

```rust
// Host side
fn host() {
    let xs = vec![92; 1024 * 1024 * 100];
    eat_stack(0);

    fn eat_stack(lvl: u32) {
        eat_stack(lvl.wrapping_add(1))
    }
}

// Wasm side
#[no_mangle]
pub fn f() -> u64 {
    unsafe { host() }
    92
}

extern "C" {
    fn host();
}
```

### Expected behavior

I would expect one of the following to happen:

* either wasmer unwinds the stack and returns an error
* or wasmer doesn't unwind the stack and aborts the process

### Actual behavior

Wasmer returns an `Err`, without cleaning up the host stack.

### Additional context

The host is `x86_64-unknown-linux-gnu`

Fun fact: the repository is polymorphic in the bug it reproduces, see https://github.com/rust-analyzer/rust-analyzer/issues/8905

**Comment by syrusakbary** [comment]
Ok. I just tested this and I was able to reproduce it.

I also just published a fix in #2342 (via `IS_WASM_PC`) and was able to run the example with the expected result (Wasmer not catching the signal but relaying it to the host):

```
calling host function
Is wasm PC: 0x1046548ec ? false

thread 'main' has overflowed its stack
fatal runtime error: stack overflow
[1]    17357 abort      cargo run --example bench
```

**Comment by syrusakbary** [comment]
The issue is now fixed on master @matklad
